File: haskell/src/Solutions/Euler686.hs
module Solutions.Euler686 where
import Data.Maybe

-- 123 * 10^k <= 2^n < 124 * 10^k
-- log_2(123) + k * log_2(10) <= n < log_2(124) + k * log_2(10)

l10 = logBase 2 10
l123 = logBase 2 123
l124 = logBase 2 124

findPower :: Double -> Maybe Int
findPower k =
    let s = k * l10
        low = ceiling $ l123 + s
        high = floor $ l124 + s
    in if low == high
        then Just high
        else Nothing

properPowers :: [Int]
properPowers = mapMaybe findPower [2.0..]

euler686 :: IO String
euler686 = return $ show $ properPowers !! 678909

File: rust/src/solutions/euler686.rs
pub fn euler686() -> String {
    let l10 = 10f64.log2();
    let l123 = 123f64.log2();
    let l124 = 124f64.log2();
    let find_power = |k: f64| {
        let s = k * l10;
        let low = (l123 + s).ceil();
        let high = (l124 + s).floor();
        if low == high {
            Some(high)
        } else {
            None
        }
    };
    (2u64..)
        .filter_map(|k| find_power(k as f64))
        .nth(678909)
        .unwrap()
        .to_string()
}
